src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -267,71 +269,175 @@ __device__ DeviceSplitCandidate operator+(const DeviceSplitCandidate& a,
   return b.loss_chg > a.loss_chg ? b : a;
 }
 
+/**
+ * \brief Set the bits for categorical splits based on the split threshold.
+ */
+template <typename GradientSumT>
+__device__ void SortBasedSplit(EvaluateSplitInputs<GradientSumT> const &input,
+                               common::Span<bst_feature_t const> d_sorted_idx, bst_feature_t fidx,
+                               bool is_left, common::Span<common::CatBitField::value_type> out,
+                               DeviceSplitCandidate *p_out_split) {
+  auto &out_split = *p_out_split;
+  out_split.split_cats = common::CatBitField{out};
+  auto node_sorted_idx =
+      is_left ? d_sorted_idx.subspan(0, input.feature_values.size())
+              : d_sorted_idx.subspan(input.feature_values.size(), input.feature_values.size());
+  size_t node_offset = is_left ? 0 : input.feature_values.size();
+  auto best_thresh = out_split.PopBestThresh();
+  auto f_sorted_idx =
+      node_sorted_idx.subspan(input.feature_segments[fidx], input.FeatureBins(fidx));
+  if (out_split.dir != kLeftDir) {
+    // forward, missing on right
+    auto beg = dh::tcbegin(f_sorted_idx);
+    // Don't put all the categories into one side
+    auto boundary = std::min(static_cast<size_t>((best_thresh + 1)), (f_sorted_idx.size() - 1));
+    boundary = std::max(boundary, static_cast<size_t>(1ul));
+    auto end = beg + boundary;
+    thrust::for_each(thrust::seq, beg, end, [&](auto c) {
```

Comment:
This is quite inefficient when number of categories is huge.  But I think it's best to push the optimization work after 1.6.